omniledger/service/service_test.go

```diff
@@ -680,6 +679,57 @@ func TestService_SetBadConfig(t *testing.T) {
 	}
 }
 
+func TestService_RotateLeader(t *testing.T) {
+	interval := 2 * time.Second
+	s := newSerN(t, 1, interval, 4, true)
+	defer s.local.CloseAll()
+	// We close only the servers that are alive, service 0 is closed to
+	// simulate failure.
+	defer func() {
+		for _, service := range s.services[1:] {
+			service.TestClose()
+		}
+	}()
+
+	for _, service := range s.services {
+		service.SetPropagationTimeout(interval / 2)
+	}
+
+	// Stop the leader, then the next node should take over.
+	s.service().TestClose()
+	s.hosts[0].Pause()
+
+	// wait for the new block
+	var ok bool
+	for i := 0; i < 5; i++ {
+		time.Sleep(2 * s.interval)
+		config, err := s.services[1].LoadConfig(s.sb.SkipChainID())
+		require.NoError(t, err)
+
+		if config.Roster.List[0].Equal(s.services[1].ServerIdentity()) {
+			ok = true
+			break
+		}
+	}
+	require.True(t, ok, "leader rotation failed")
+
+	// check the leader for all nodes
+	for _, service := range s.services[1:] {
+		// everyone should have the same leader after the genesis block is stored
+		leader, err := service.getLeader(s.sb.SkipChainID())
+		require.NoError(t, err)
+		require.NotNil(t, leader)
+		require.True(t, leader.Equal(s.services[1].ServerIdentity()))
+	}
+
+	// try to send a transaction to the node on index 2 (not the current leader)
```

Comment:
not or now? So if I understand this, this is a test of "server 0 was leader, now is down, server 1 is now leader, so send tx to server 2 and see that server 1 polls server 2, finds the tx, and that we can talk to server 2 and see the tx applied".

This is an excellent end to end test, and needs better comment to point that out.

Also, it would be more interesting to ask server 3 for the inclusion proof, showing that the tx was propagated from the new leader to a follower who never heard anything about the tx.

Reply:
you're understanding is spot-on. I will add comments to make the test understandable and do the inclusion proof for index 3.